webrtc/README.md

```diff
@@ -0,0 +1,268 @@
+    - [Multiplexing](#multiplexing)
+    - [General Open Questions](#general-open-questions)
+    - [Previous, ongoing and related work](#previous-ongoing-and-related-work)
+- [FAQ](#faq)
+
+<!-- markdown-toc end -->
+
+
+## Motivation
+
+1. **No need for valid TLS certificates.** Enable browsers to connect to public
+   server nodes without those server nodes providing a TLS certificate within
+   the browsers trustchain. Note that we can not do this today with our
+   Websocket transport.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   non-public server node to connect.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Addressing
+
+WebRTC multiaddresses are composed of an IP and UDP address component, followed
+by `/webrtc` and a multihash of the certificate that the node uses.
+
+Examples:
+- `/ip4/1.2.3.4/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+- `/ip6/fe80::1ff:fe23:4567:890a/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`
+
+## Connection Establishment
+
+### Browser to public Server
+
+Scenario: Browser _A_ wants to connect to server node _B_ where _B_ is publicly
+reachable but _B_ does not have a TLS certificate trusted by _A_.
+
+As a preparation browser _A_ [generates a
+certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate)
+and [gets the certificate's
+fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`),
+  through some external mechanism.
+
+2. _A_ instantiates a `RTCPeerConnection`, passing its local certificate as a
+   parameter. See
+   [`RTCPeerConnection()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection).
+
+3. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr and sets it
+   via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+4. _A_ establishes the connection to _B_. _B_ learns of _A_ TLS fingerprint
+   through A's STUN message USERNAME attribute. See Open Questions below for
+   potential better solutions.
+
+5. See [Connection Security](#connection-security).
+
+6. See [Multiplexing](#multiplexing).
+
+#### Open Questions
+
+- How does the server learn the TLS certificate fingerprint of the browser? Is
+  embedding A's TLS certificate fingerprint in A's STUN message USERNAME
+  attribute the best option?
+
+- Is the fact that the server accepts STUN messages from the client prone to
+  attacks?
+
+    - Can an attacker launch an **amplification attack** with the STUN endpoint
+      of the server?
```

Comment:
I think this is answered by the STUN RFC in https://datatracker.ietf.org/doc/html/rfc5389#section-16.1.2: there's no amplification in terms of number of packets, and slight amplification in terms of packet size.

When thinking about amplification, it's useful to consider what QUIC does, especially since there is a huge fleet of go-libp2p nodes out there that exposes QUIC endpoints: In QUIC, the server is allowed to send 3x the number of bytes it has received, after processing the Initial packet from the client (which will result in 3 packets being sent).

I'd like to get a better feeling for the payload / size of STUN packets we'll be sending and receiving, but my hunch is that using QUIC will be the more interesting amplification vector.

Reply:
Documented in f5eb1b7. I will resolve here. The "Open Question" will make sure we will get to it once more.